[SCEVExpander] Relax hoisting condition for AddRec start

averine opened the pull request:
Don't require the existence of a preheader to hoist an AddRec start value: a unique loop predecessor which isLegalToHoistInto should be enough.

Commit:
[SCEVExpander] Relax hoisting condition for AddRec start

Don't require the existence of a preheader to hoist an
AddRec start value: a unique loop predecessor which
isLegalToHoistInto should be enough.

averine commented:
Suggested reviewers: @preames  @sanjoy.

sanjoy commented on llvm/lib/Transforms/Utils/ScalarEvolutionExpander.cpp:
```diff
@@ -980,11 +980,14 @@ SCEVExpander::getAddRecExprPHILiterally(const SCEVAddRecExpr *Normalized,
-      expand(Normalized->getStart(), L->getLoopPreheader()->getTerminator());
+  // Expand code for the start value into the loop predecessor. The loop is not
+  // necessarily in Loop Simplify Form, so assert it's legal to do so.
+  auto *LP = L->getLoopPredecessor();
```
Is `LP` guaranteed to dominate the loop body?

averine replied:
No: At that line 985, I cannot guarantee that `LP` dominates the loop body.
This is "checked" this in the `assert` line 986 (by definition of `getLoopPredecessor()`, `LP` is the unique predecessor of the header iff it's non-zero).

So the patch helps when `L` has no preheader but a unique predecessor (which is legal to hoist into), but leads to the same error when `L` has several predecessors.

sanjoy replied:
Sounds good.

sanjoy commented:
Thanks, I don't have any objections on the PR, but please wait on an explicit approval from someone more current on the codebase (@nikic or @preames).

nikic commented:
I don't think this fix is right. There is a check in isSafeToExpand() that makes sure there is a pre-header for addrecs in non-canonical mode. If you hit this assertion failure, it means that an isSafeToExpand() check is missing somewhere.

preames commented on llvm/lib/Transforms/Utils/ScalarEvolutionExpander.cpp:
```diff
@@ -980,11 +980,14 @@ SCEVExpander::getAddRecExprPHILiterally(const SCEVAddRecExpr *Normalized,
-         "Can't expand add recurrences without a loop preheader!");
-  Value *StartV =
-      expand(Normalized->getStart(), L->getLoopPreheader()->getTerminator());
+  // Expand code for the start value into the loop predecessor. The loop is not
```
This looks suspect.  The difference between a preheader and a predecessor is whether the loop header is guaranteed to execute after the execution of the former.  There's two concerns here:
1) This ends up speculating instructions onto paths which didn't previous execute them.  Might be legally fine, but profitability is suspect.  
2) I believe expanded sometimes uses "must execute" reasoning to prove safety (i.e. non speculation), such logic would become unsound with this change (I think? - haven't fully traced it through.)

I agree with @nikic that this looks like a isSafeToExpand check missing somewhere.

Commit:
Revert "[SCEVExpander] Relax hoisting condition for AddRec start"

This reverts commit 3350c5ec136b406a717d00b2482ddf8cba478e6d.

Commit:
Address reviewers' comments: add a check before expansion

This canExpand() check verifies addrecs have a preheader
in non-canonical mode.

averine commented:
> I don't think this fix is right. There is a check in isSafeToExpand() that makes sure there is a pre-header for addrecs in non-canonical mode. If you hit this assertion failure, it means that an isSafeToExpand() check is missing somewhere.

Thanks for pointing this out. I added such a check before SCEV expansion is attempted (I did not use isSafeToExpand() but a weaker version of it which just checks the presence of the preheader, else indvars pass would not hoist some divisions in unit tests).

nikic commented:
@averine Can you share the test diffs if you just use isSafeToExpand()? IndVars hoisting divisions sounds fishy...

nikic commented on llvm/test/Transforms/IndVarSimplify/scevexpander-in-non-simplified-loop.ll:
```diff
@@ -0,0 +1,39 @@
+; RUN: opt --passes=indvars -disable-output < %s
```
Use update_test_checks.py.

averine commented:
@nikic here are the tests failing when I replace canExpand() by isSafeToExpand():
```
  LLVM :: Transforms/IndVarSimplify/X86/pr35406.ll
  LLVM :: Transforms/IndVarSimplify/pr30806-phi-scev.ll
  LLVM :: Transforms/IndVarSimplify/pr64891.ll
  LLVM :: Transforms/LoopVectorize/pr38697.ll
```
For these tests, diffs of the output of the "opt -passes=indvars" line are below.
Having a second look there is no diff in division hoisting.

```
--- IndVarSimplify/pr30806-phi-scev.ll.canExpand.opt	2024-01-02 17:55:43
+++ IndVarSimplify/pr30806-phi-scev.ll.isSafeToExpand.opt	2024-01-02 17:55:03
@@ -20 +19,0 @@
-  %indvars.iv = phi i64 [ %indvars.iv.next, %while.body ], [ 0, %while.body.lr.ph ]
@@ -22,5 +21,6 @@
-  %2 = sext i32 %div to i64
-  %indvars.iv.next = add i64 %indvars.iv, %2
-  %3 = load i32, ptr @theSize, align 4
-  store i32 %3, ptr %i, align 4
-  call void @bar(ptr nonnull %i, i64 %indvars.iv.next)
+  %inx.06 = phi i32 [ 0, %while.body.lr.ph ], [ %add, %while.body ]
+  %add = add nsw i32 %inx.06, %div
+  %2 = load i32, ptr @theSize, align 4
+  store i32 %2, ptr %i, align 4
+  %conv = sext i32 %add to i64
+  call void @bar(ptr nonnull %i, i64 %conv)
@@ -28,2 +28,2 @@
-  %4 = load i32, ptr %flag, align 4
-  %tobool = icmp eq i32 %4, 0
+  %3 = load i32, ptr %flag, align 4
+  %tobool = icmp eq i32 %3, 0
--- IndVarSimplify/X86/pr35406.ll.canExpand.opt	2024-01-02 17:22:33
+++ IndVarSimplify/X86/pr35406.ll.isSafeToExpand.opt	2024-01-02 17:23:57
@@ -11,2 +11,2 @@
-  %indvars.iv = phi i64 [ %indvars.iv.next, %loop2.exit ], [ 8, %entry ]
-  %exitcond = icmp eq i64 %indvars.iv, 15
+  %local_0_ = phi i32 [ 8, %entry ], [ %i9, %loop2.exit ]
+  %exitcond = icmp eq i32 %local_0_, 15
@@ -31 +31 @@
-  %indvars.iv.next = add nuw nsw i64 %indvars.iv, 1
+  %i9 = add nuw nsw i32 %local_0_, 1
--- LoopVectorize/pr38697.ll.canExpand.opt	2024-01-02 17:22:33
+++ LoopVectorize/pr38697.ll.isSafeToExpand.opt	2024-01-02 17:23:57
@@ -111 +111 @@
-  %0 = zext i32 %start1 to i64
+  %umax = call i32 @llvm.umax.i32(i32 %start1, i32 100)
@@ -115 +114,0 @@
-  %indvars.iv = phi i64 [ %indvars.iv.next, %for.end ], [ %0, %entry ]
@@ -117,3 +116,3 @@
-  %indvars3 = trunc i64 %indvars.iv to i32
-  %cmp = icmp ult i32 %indvars3, 100
-  br i1 %cmp, label %for.body, label %for.end10
+  %counter1.0 = phi i32 [ %start1, %entry ], [ %inc9, %for.end ]
+  %exitcond1 = icmp ne i32 %counter1.0, %umax
+  br i1 %exitcond1, label %for.body, label %for.end10
@@ -131 +130 @@
-  %indvars.iv1 = phi i64 [ %indvars.iv.next2, %for.body3 ], [ 0, %for.body3.lr.ph ]
+  %index.027 = phi i32 [ 0, %for.body3.lr.ph ], [ %add4, %for.body3 ]
@@ -134,3 +133,4 @@
-  %1 = udiv i64 16, %indvars.iv
-  %indvars.iv.next2 = add nuw nsw i64 %indvars.iv1, %1
-  %arrayidx6 = getelementptr inbounds i32, ptr %ptr, i64 %indvars.iv.next2
+  %div = udiv i32 16, %counter1.0
+  %add4 = add nuw nsw i32 %div, %index.027
+  %idxprom5 = zext i32 %add4 to i64
+  %arrayidx6 = getelementptr inbounds i32, ptr %ptr, i64 %idxprom5
@@ -149 +149 @@
-  %indvars.iv.next = add nuw nsw i64 %indvars.iv, 1
+  %inc9 = add i32 %counter1.0, 1
@@ -252,0 +253,3 @@
+; Function Attrs: nocallback nofree nosync nounwind speculatable willreturn memory(none)
+declare i32 @llvm.umax.i32(i32, i32) #1
+
@@ -253,0 +257 @@
+attributes #1 = { nocallback nofree nosync nounwind speculatable willreturn memory(none) }
--- IndVarSimplify/pr64891.ll.canExpand.opt	2024-01-02 17:22:33
+++ IndVarSimplify/pr64891.ll.isSafeToExpand.opt	2024-01-02 17:23:57
@@ -6,0 +7,2 @@
+  %div = sdiv i32 1, 0
+  %trunc = trunc i32 %div to i16
@@ -9,0 +12,4 @@
+  %phi = phi i16 [ 0, %entry ], [ %or, %loop ]
+  %or = or disjoint i16 %phi, %trunc
+  %phi.ext = sext i16 %phi to i64
+  %add.ptr = getelementptr i8, ptr null, i64 %phi.ext
```

Commit:
Address reviewer's comment: add autogenerated CHECKs

averine commented:
Ping

averine commented:
Ping

averine commented:
Ping

nikic commented:
I don't think introducing canExpand() is right, we should be using isSafeToExpand(). But it likely requires additional changes to avoid regressions (something like consider division safe to expand if it's actually going to be reused from an existing value). I didn't have a chance to try that yet.
